### PR TITLE
8292250: Create test for co-located JDI MethodEntry, Step, and Breakpoint events

### PR DESCRIPTION
**Pull request by plummercj**
We currently have no tests for co-located MethodEntry, Step, and Breakpoint events. We should make sure they are being properly co-located as described in the JDI spec, and also do special test cases for [JDK-8292217](https://bugs.openjdk.org/browse/JDK-8292217).

https://docs.oracle.com/en/java/javase/17/docs/api/jdk.jdi/com/sun/jdi/event/EventSet.html

And sorry in advance that the logic is a bit hard to follow in this test due to having multiple test cases, and dealing with the async nature of JDI testing. All I can say is that is used to be a lot worse before I did multiple passes to improve it.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292250](https://bugs.openjdk.org/browse/JDK-8292250): Create test for co-located JDI MethodEntry, Step, and Breakpoint events

### Reviewers
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9840/head:pull/9840` \
`$ git checkout pull/9840`

Update a local copy of the PR: \
`$ git checkout pull/9840` \
`$ git pull https://git.openjdk.org/jdk pull/9840/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9840`

View PR using the GUI difftool: \
`$ git pr show -t 9840`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9840.diff">https://git.openjdk.org/jdk/pull/9840.diff</a>

</details>

### CODE REVIEW

**Comment by plummercj** [comment]
Here is output from a test run that might be useful when reviewing the test:

Got main thread: instance of java.lang.Thread(name='main', id=1)
Waiting for events: 

EventSet for test case #0: event set, policy:2, count:1 = {BreakpointEvent@CLEDebugee:75 in thread main}
Got BreakpointEvent(1): CLEDebugee.test1:75

EventSet for test case #1: event set, policy:2, count:5 = {MethodEntryEvent@java.lang.ClassLoader:521 in thread main, MethodEntryEvent@jdk.internal.loader.ClassLoaders$AppClassLoader:180 in thread main, MethodEntryEvent@java.lang.System:497 in thread main, MethodEntryEvent@java.lang.System:205 in thread main, MethodExitEvent@java.lang.System:205 in thread main}
TESTCASE #1 FAILED: too many events in EventSest: 5
Got MethodEntryEvent: java.lang.ClassLoader.loadClass:521
Got MethodEntryEvent: jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass:180
Got MethodEntryEvent: java.lang.System.getSecurityManager:497
Got MethodEntryEvent: java.lang.System.allowSecurityManager:205
Got MethodExitEvent: java.lang.System.allowSecurityManager:205

EventSet for test case #1: event set, policy:2, count:1 = {StepEvent@CLEDebugee:76 in thread main}
Got StepEvent: CLEDebugee.test1:76

EventSet for test case #1: event set, policy:2, count:1 = {BreakpointEvent@CLEDebugee:78 in thread main}
Got BreakpointEvent(2): CLEDebugee.test2:78

EventSet for test case #2: event set, policy:2, count:5 = {MethodEntryEvent@java.lang.ClassLoader:521 in thread main, MethodEntryEvent@jdk.internal.loader.ClassLoaders$AppClassLoader:180 in thread main, MethodEntryEvent@java.lang.System:497 in thread main, MethodEntryEvent@java.lang.System:205 in thread main, MethodExitEvent@java.lang.System:205 in thread main}
TESTCASE #2 FAILED: too many events in EventSest: 5
Got MethodEntryEvent: java.lang.ClassLoader.loadClass:521
Got MethodEntryEvent: jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass:180
Got MethodEntryEvent: java.lang.System.getSecurityManager:497
Got MethodEntryEvent: java.lang.System.allowSecurityManager:205
Got MethodExitEvent: java.lang.System.allowSecurityManager:205

EventSet for test case #2: event set, policy:2, count:1 = {StepEvent@t2:44 in thread main}
Got StepEvent: t2.foo:44

EventSet for test case #2: event set, policy:2, count:1 = {BreakpointEvent@CLEDebugee:63 in thread main}
Got BreakpointEvent(3): CLEDebugee.runTests:63

EventSet for test case #3: event set, policy:2, count:3 = {MethodEntryEvent@CLEDebugee:87 in thread main, StepEvent@CLEDebugee:87 in thread main, BreakpointEvent@CLEDebugee:87 in thread main}
Got MethodEntryEvent: CLEDebugee.test3:87
Got StepEvent: CLEDebugee.test3:87
Got BreakpointEvent(4): CLEDebugee.test3:87
TESTCASE #3 PASSED

EventSet for test case #3: event set, policy:2, count:1 = {BreakpointEvent@CLEDebugee:64 in thread main}
Got BreakpointEvent(5): CLEDebugee.runTests:64

EventSet for test case #4: event set, policy:2, count:2 = {MethodEntryEvent@CLEDebugee:95 in thread main, BreakpointEvent@CLEDebugee:95 in thread main}
Got MethodEntryEvent: CLEDebugee.test4:95
Got BreakpointEvent(6): CLEDebugee.test4:95
TESTCASE #4 PASSED

EventSet for test case #4: event set, policy:2, count:1 = {BreakpointEvent@CLEDebugee:65 in thread main}
Got BreakpointEvent(7): CLEDebugee.runTests:65

EventSet for test case #5: event set, policy:2, count:2 = {StepEvent@CLEDebugee:102 in thread main, BreakpointEvent@CLEDebugee:102 in thread main}
Got StepEvent: CLEDebugee.test5:102
Got BreakpointEvent(8): CLEDebugee.test5:102
TESTCASE #5 PASSED

EventSet for test case #5: event set, policy:2, count:1 = {BreakpointEvent@CLEDebugee:66 in thread main}
Got BreakpointEvent(9): CLEDebugee.runTests:66

EventSet for test case #6: event set, policy:2, count:2 = {MethodEntryEvent@CLEDebugee:109 in thread main, StepEvent@CLEDebugee:109 in thread main}
Got MethodEntryEvent: CLEDebugee.test6:109
Got StepEvent: CLEDebugee.test6:109
TESTCASE #6 PASSED

EventSet for test case #6: event set, policy:2, count:2 = {VMDeathEvent, VMDeathEvent}
All done...

EventSet for test case #6: event set, policy:0, count:1 = {VMDisconnectEvent}

**Comment by plummercj** [comment]
Ping!

File: test/jdk/com/sun/jdi/CLETest.java [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+        BreakpointEvent bpe = startToMain("CLEDebugee");
+        targetClass = (ClassType)bpe.location().declaringType();
+        mainThread = bpe.thread();
+        System.out.println("Got main thread: " + mainThread);
+        erm = eventRequestManager();
+
+        try {
+            // Setup all breakpoints
+            for (MethodBreakpointData bpData : breakpoints) {
+                Location loc = findMethodLocation(targetClass, bpData.method,
```
Comment (alexmenkov) [inline test/jdk/com/sun/jdi/CLETest.java:409]: Specify breakpoints by line number in a method is better than just line number, but did you think about mark them with some tag and parse source file to get line numbers? (like in test/jdk/com/sun/jdi/lib/jdb/JdbTest.java, parseBreakpoints method)

Comment (plummercj) [reply test/jdk/com/sun/jdi/CLETest.java:413]: It looks like the tests in com/sun/jdi are split between those that extend `JdbTest` and those that extend `TestScaffold`. I'm not too sure of the rationale for using one over the other, but I ended up with `TestScaffold` because of the test I initially used as a template (BreakpointTest.java). `parseBreakpoints()` is only supported by `JdbTest`, so I would need to convert the test to extend it instead. I have no idea how disruptive this might be to the test.

I think the breakpoints I have are fairly safe from code movement and modifications. It would require edits within the methods with the breakpoints, all of which are very simple and well commented, and there really is no reason to ever edit them in the first place. Also there are 8 other tests that are sensitive to line numbers (search for THIS TEST IS LINE NUMBER SENSITIVE), and they are at much greater risk (adding or removing a line anywhere before the breakpoint will break the test).

Comment (alexmenkov) [reply test/jdk/com/sun/jdi/CLETest.java:413]: AFAIR JdbTest is a base class to test jdb functionality, I didn't mean to extend it. But breakpoint parsing code is quite simple, I thought about add similar method to TestScaffold instead of  findMethodLocation.
As I wrote your way with findMethodLocation is much better than just line numbers (used in "THIS TEST IS LINE NUMBER SENSITIVE" tests), I'm okay with it.

**Review by alexmenkov — APPROVED** [review]

File: test/jdk/com/sun/jdi/CLETest.java (hunk 1 above)
Comment (plummercj) [reply test/jdk/com/sun/jdi/CLETest.java:413]: Ok. Thanks for the review!

**Review by kevinjwalls — APPROVED** [review]

**Review by kevinjwalls — APPROVED** [review]

**Comment by plummercj** [comment]
I had to undo the last change that checked if the breakpoint was the expected one. There were two issues. The first was that the breakpoints are not in order, so it always threw an exception on the first breakpoint. However, the test still passed. That was because of a pre-existing bug. I declared `testFailed` locally within the test class. The overrode the one declared in `TestScaffold`, and which gets set `false` if there is an exception thrown in the event handler. I got rid of the override, and then test started to properly fail due to the incorrect breakpoint checks, which I have now also removed.

**Review by kevinjwalls — APPROVED** [review]
OK thanks, the extra breakpoint check might have been nice but maybe not worth it right now.

**Review by alexmenkov — APPROVED** [review]
Still LGTM

**Comment by plummercj** [comment]
Thanks!
